### ISSUE TITLE
Add declaration-property-max-values

### ISSUE THREAD

**Issue by keganlance** [stylelint/stylelint#5435]
Since the padding and margin shorthands are hard to understand for many people, it can be preferred to write it out fully.

Consider the following scenario:
margin: 2px 8px 3px;

Most people have no idea what this involves. Going for a syntax like this:
margin: 8px;
margin-top: 2px;
margin-bottom: 3px;

Would be preferred.

Rule could be to limit the amount of values padding, margin and border-width can have.

**Comment by jeddy3** [comment]
@keganlance Thanks for the second request.

This has [come up before](https://github.com/stylelint/stylelint/issues/3968), and [was considered a good candidate for a plugin](https://github.com/stylelint/stylelint/issues/3968#issuecomment-468619412). However, I'm less convinced it's a problematic rule now _if we make it per property_, e.g.:

```json
{ 
  "rules": {
    "declaration-property-max-values": {
      "margin": 2,
      "gap": 1
      ..
     }
   }
}
```

There are [quite a few (shorthand) properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Shorthand_properties#see_also) that could make use of this rule.

We can add a `ignoreValues: []` secondary option for things like the [`logical` keyword](https://drafts.csswg.org/css-logical/#logical-shorthand-keyword).

I'll label as "discussion" to see if anyone else has thoughts on this. In the meantime, you're welcome to implement this rule as [a plugin](https://stylelint.io/developer-guide/plugins).

---

Incidentally, the problem can somewhat be solved by adopting logical properties, where you'd rewrite the following:

```css
a { 
  margin: 2px 8px 3px;
}
```

As:

```css
a {
  margin-block: 8px;
  margin-inline: 2px 3px;
}
```

These properties are limited to 1 or two values, and it's clearer that two values refer to _start_ and _end_. You'd probably want to disallow the `margin` shorthand using the [`property-disallowed-list` rule](https://stylelint.io/user-guide/rules/list/property-disallowed-list/). Having said that, I imagine some users would prefer a single property for uniform margins, and so a `declaration-property-max-values` rule would be useful here too:

```json
{ 
  "rules": {
    "declaration-property-max-values": [
      {
        "margin": 1
       }, { 
        "ignoreValues": ["logical"] 
      } 
     ]
   }
}
```

Which would allow:

```css
a {
  margin: logical 2px;
}
```

**Comment by jeddy3** [comment]
As there are no objections to this rule, I'll label as good to go:

- Name: `declaration-property-max-values`
- Primary option: `object - { regex/string: int }`
- Secondary options: None
- Autofixable: No
- Message: `Expected "${property}" to have no more than ${max} ${max === 1 ? 'value' : 'values'}`
- Description: "Limit the number of values for a list of properties within declarations."
- Section: "Limit language features" -> "Declarations"

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

**Comment by mattxwang** [comment]
Just peeking through some issues - do you think the rule's strategy will be similar to [declaration-property-value-allowed-list](https://github.com/stylelint/stylelint/tree/main/lib/rules/declaration-property-value-allowed-list), but checking the length of the property values array instead of matching the value against a regex? If so, I can take a crack at implementing this sometime soon!

**Comment by jeddy3** [comment]
Yes, it'll have a similar shape to `declaration-property-value-allowed-list`.

**Comment by mattxwang** [comment]
Hey @keganlance, this new rule just got merged in [on version 14.6.0](https://github.com/stylelint/stylelint/releases/tag/14.6.0). Let me know if this fulfills your original use-case, even if it's been a bit after the original issue was filed!
